base/applications/kbswitch/kbswitch.c

```diff
@@ -552,34 +619,20 @@ DeleteHooks(VOID)
 ULONG
 GetNextLayout(VOID)
 {
-    TCHAR szLayoutNum[3 + 1], szLayoutID[CCH_LAYOUT_ID + 1];
-    ULONG uLayoutNum, uMaxNum = GetMaxLayoutNum();
-
-    for (uLayoutNum = ulCurrentLayoutNum + 1; ; ++uLayoutNum)
-    {
-        if (uLayoutNum > uMaxNum)
-            uLayoutNum = 1;
-        if (uLayoutNum == ulCurrentLayoutNum)
-            break;
-
-        _ultot(uLayoutNum, szLayoutNum, 10);
-        if (GetLayoutID(szLayoutNum, szLayoutID, ARRAYSIZE(szLayoutID)))
-            return uLayoutNum;
-    }
-
-    return ulCurrentLayoutNum;
+    UINT uMaxNum = GetMaxLayoutNum();
```

Comment:
Shouldn't you use actually the currently cached `g_cKLs` value that you retrieve in many other places with the repeated `GetKeyboardLayoutList` function call? -- It would appear otherwise that you could get a situation where what GetNextLayout() returns gets unsynchronized with the other cached values, especially if some new keyboard layouts get added to, or removed from the system...

Reply:
Added `UpdateLayoutList` helper function that uses `GetKeyboardLayoutList`.